File: packages/p2p/src/wire.js
export const MAGIC = 0x5c;

export const FRAME = Object.freeze({
  REQUEST: 0x01,
  DATA: 0x02,
  DATA_END: 0x03,
  CANCEL: 0x04,
  BITFIELD: 0x05,
  REJECT: 0x06,
  CODED: 0x07,
  RANK: 0x08
});

export const REJECT_REASON = Object.freeze({
  DONT_HAVE: 1,
  BUSY: 2,
  QUOTA: 3
});

export function frame(type, seq, payload = Buffer.alloc(0)) {
  const body = Buffer.from(payload);
  const out = Buffer.alloc(10 + body.length);
  out[0] = MAGIC;
  out[1] = type;
  out.writeUInt32BE(seq >>> 0, 2);
  out.writeUInt32BE(body.length, 6);
  body.copy(out, 10);
  return out;
}

export function parseFrame(buf) {
  const input = Buffer.from(buf);
  if (input.length < 10 || input[0] !== MAGIC) return null;
  const type = input[1];
  const seq = input.readUInt32BE(2);
  const length = input.readUInt32BE(6);
  if (length !== input.length - 10) return null;
  return {
    type,
    seq,
    payload: input.subarray(10)
  };
}

export function bitfieldPayload(seqs) {
  const values = [...seqs].map((seq) => seq >>> 0);
  const out = Buffer.alloc(values.length * 4);
  values.forEach((seq, index) => out.writeUInt32BE(seq, index * 4));
  return out;
}

export function parseBitfieldPayload(payload) {
  const input = Buffer.from(payload);
  if (input.length % 4 !== 0) return null;
  const out = new Set();
  for (let offset = 0; offset < input.length; offset += 4) {
    out.add(input.readUInt32BE(offset));
  }
  return out;
}

export function codedPayload(packet) {
  if (packet.coeffs.length > 255) throw new Error("coefficient vector too large");
  const coeffs = Buffer.from(packet.coeffs);
  const data = Buffer.from(packet.data);
  return Buffer.concat([Buffer.from([coeffs.length]), coeffs, data]);
}

export function parseCodedPayload(payload) {
  const input = Buffer.from(payload);
  if (input.length < 1) return null;
  const k = input[0];
  if (input.length < 1 + k) return null;
  return {
    coeffs: input.subarray(1, 1 + k),
    data: input.subarray(1 + k)
  };
}

export function rankPayload(seq, rank) {
  const out = Buffer.alloc(6);
  out.writeUInt32BE(seq >>> 0, 0);
  out.writeUInt16BE(rank & 0xffff, 4);
  return out;
}

export function parseRankPayload(payload) {
  const input = Buffer.from(payload);
  if (input.length !== 6) return null;
  return {
    seq: input.readUInt32BE(0),
    rank: input.readUInt16BE(4)
  };
}

File: android/app/src/main/java/tv/swarmcast/p2p/Wire.kt
package tv.swarmcast.p2p

import java.nio.ByteBuffer

object Wire {
    const val MAGIC: Byte = 0x5C
    const val CHUNK: Int = 16 * 1024
    const val REQUEST: Byte = 0x01
    const val DATA: Byte = 0x02
    const val DATA_END: Byte = 0x03
    const val CANCEL: Byte = 0x04
    const val BITFIELD: Byte = 0x05
    const val REJECT: Byte = 0x06
    const val CODED: Byte = 0x07
    const val RANK: Byte = 0x08
    const val CODED_REQUEST: Byte = 0x09

    fun frame(type: Byte, seq: Int, payload: ByteArray = ByteArray(0)): ByteBuffer =
        ByteBuffer.allocate(10 + payload.size).apply {
            put(MAGIC)
            put(type)
            putInt(seq)
            putInt(payload.size)
            put(payload)
            flip()
        }

    data class Msg(val type: Byte, val seq: Int, val payload: ByteArray)
    data class CodedPayload(val coeffs: ByteArray, val data: ByteArray)
    data class RankPayload(val seq: Int, val rank: Int)

    fun parse(buf: ByteBuffer): Msg? {
        if (buf.remaining() < 10 || buf.get() != MAGIC) return null
        val type = buf.get()
        val seq = buf.int
        val len = buf.int
        if (len != buf.remaining()) return null
        val payload = ByteArray(len)
        buf.get(payload)
        return Msg(type, seq, payload)
    }

    fun bitfield(seqs: Collection<Int>): ByteArray =
        ByteBuffer.allocate(seqs.size * 4).apply {
            seqs.forEach { putInt(it) }
            flip()
        }.let { buffer -> ByteArray(buffer.remaining()).also { buffer.get(it) } }

    fun parseBitfield(payload: ByteArray): Set<Int>? {
        if (payload.size % 4 != 0) return null
        val buffer = ByteBuffer.wrap(payload)
        val out = LinkedHashSet<Int>()
        while (buffer.remaining() >= 4) out.add(buffer.int)
        return out
    }

    fun codedPayload(coeffs: ByteArray, data: ByteArray): ByteArray {
        require(coeffs.size <= 255) { "coefficient vector too large" }
        return ByteBuffer.allocate(1 + coeffs.size + data.size).apply {
            put(coeffs.size.toByte())
            put(coeffs)
            put(data)
        }.array()
    }

    fun parseCodedPayload(payload: ByteArray): CodedPayload? {
        if (payload.isEmpty()) return null
        val coeffCount = payload[0].toInt() and 0xff
        if (payload.size < 1 + coeffCount) return null
        return CodedPayload(
            coeffs = payload.copyOfRange(1, 1 + coeffCount),
            data = payload.copyOfRange(1 + coeffCount, payload.size)
        )
    }

    fun rankPayload(seq: Int, rank: Int): ByteArray =
        ByteBuffer.allocate(6).apply {
            putInt(seq)
            putShort((rank and 0xffff).toShort())
        }.array()

    fun parseRankPayload(payload: ByteArray): RankPayload? {
        if (payload.size != 6) return null
        val buffer = ByteBuffer.wrap(payload)
        return RankPayload(
            seq = buffer.int,
            rank = buffer.short.toInt() and 0xffff
        )
    }
}
